python/shotgun_fields/bubble_widget.py

```diff
@@ -0,0 +1,495 @@
+            lambda: self.remove_clicked.emit()
+        )
+
+    def set_text(self, label_text):
+        """
+        Set the bubble's display text.
+
+        :param str label_text: The display text
+        """
+        self.text_label.setText(label_text)
+
+    def set_image(self, url):
+        """
+        Set the bubble's display image.
+
+        :param str url:  The image url to display in the bubble.
+        """
+        self.image_label.setText("<img src='%s'/>" % url)
+
+    def set_removable(self, removable):
+        """
+        Set whether or not the bubble is removable.
+
+        Shows or hides the ``x`` button depending on the value of the
+        ``removable`` argument.
+        :param bool removable: ``True`` if the bubble is removable, ``False`` otherwise.
+        """
+        if removable:
+            self.remove_button.show()
+        else:
+            self.remove_button.hide()
+
+    def get_data(self):
+        """
+        Returns the underlying data object this widget represents.
+
+        The return type is intentionally unspecified since, in theory, the
+        bubble could represent any type of data.
+        """
+        return self._data
+
+    def set_data(self, data):
+        """
+        Set the underlying data object that this widget represents.
+
+        The type of ``data`` is intentionally unspecified since, in theory, the
+        bubble could represent any type of data.
+        """
+        self._data = data
+
+
+class BubbleEditWidget(QtGui.QTextEdit):
+    """
+    This is a base class for "bubble" entry widgets.
+
+    Each object is represented by a "bubble" similar to email address entry
+    widgets in modern email clients. Subclasses will typically handle the user
+    interaction and decide when a new "bubble" should be added to the widget.
+    """
+
+    # used as a placeholder for a bubble in the editor
+    _OBJECT_REPLACEMENT_CHAR = unichr(0xfffc)
+
+    def __init__(self, parent=None):
+        """
+        Initialize the widget.
+
+        :param parent: This widget's parent widget
+        :type parent: :class:`~PySide.QtGui.QWidget`
+        """
+
+        super(BubbleEditWidget, self).__init__(parent)
+
+        self._formats = {}
+
+        self._bubble_text_object = _BubbleTextObject(self)
+        self.document().documentLayout().registerHandler(
+            _BubbleTextObject.OBJECT_TYPE, self._bubble_text_object,
+        )
+
+        self.setMouseTracking(True)
+        self.viewport().installEventFilter(self)
+
+    def add_bubble(self, bubble):
+        """
+        Add the supplied :class:`.BubbleWidget` instance to the editor.
+
+        :param bubble: The bubble widget instance.
+        :return: A unique id for the added bubble
+        :rtype: :obj:`int`
+        """
+
+        bubble_id = self._bubble_text_object.add_bubble(bubble)
+
+        # create a character format for the bubble
+        format = QtGui.QTextCharFormat()
```

Comment:
`format` is a built-in function, so I would suggest a different variable name here